### PR TITLE
Add Swagger spec of MICROSOFT.DASHBOARD for Public Preview

### PR DESCRIPTION
**Pull request by jinghzhu-msft**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [x] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that label "WaitForARMFeedback" will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json [hunk 1]
```diff
@@ -0,0 +1,748 @@
...
+          "type": "string"
+        },
+        "target": {
+          "type": "string"
+        },
+        "details": {
+          "type": "array",
+          "x-ms-identifiers": [],
+          "items": {
+            "$ref": "#/definitions/ErrorResponseBody"
```
Comment (raosuhas) [inline specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json:378]: >ErrorResponseBody [](http://example.com/codeflow?start=35&length=17)

Consider using this from the common-types : 

[https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v3/types.json#L331](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v3/types.json#L331)

File: specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json [hunk 2]
```diff
@@ -0,0 +1,748 @@
...
+          "readOnly": true
+        },
+        "name": {
+          "description": "Name of the grafana resource.",
+          "type": "string",
+          "readOnly": true
+        },
+        "type": {
+          "description": "The type of the grafana resource.",
+          "type": "string",
```
Comment (raosuhas) [inline specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json:400]: Refer to common types and tracked resource for this : 
[https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v3/types.json#L56](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v3/types.json#L56)

File: specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json [hunk 3]
```diff
@@ -0,0 +1,748 @@
...
+          "type": "object",
+          "description": "The new tags of the grafana resource.",
+          "additionalProperties": {
+            "type": "string"
+          }
+        },
+        "location": {
+          "description": "The location of the grafana resource.",
+          "type": "string"
+        },
```
Comment (raosuhas) [inline specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json:494]: you cannot update the location in a PATCH command

File: specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json [hunk 4]
```diff
@@ -0,0 +1,748 @@
...
+        "ManagedIdentity",
+        "Key"
+      ],
+      "type": "string",
+      "x-ms-enum": {
+        "modelAsString": true,
+        "name": "LastModifiedByType"
+      }
+    },
+    "ManagedIdentity": {
```
Comment (raosuhas) [inline specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json:514]: Use common types as explained above

File: specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json [hunk 5]
```diff
@@ -0,0 +1,748 @@
...
+          "description": "Operation name, i.e., {provider}/{resource}/{operation}.",
+          "type": "string"
+        },
+        "isDataAction": {
+          "description": "Indicates whether the operation applies to data-plane. Set \"true\" for data-plane operations and \"false\" for ARM/control-plane operations.",
+          "type": "boolean"
+        },
+        "display": {
+          "$ref": "#/definitions/OperationDisplay"
+        }
```
Comment (raosuhas) [inline specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json:606]: Again Commontypes

File: specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json (hunk 4 above)
Comment (jinghzhu-msft) [reply specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json:504]: Hi @raosuhas , after comparing it with https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v3/types.json#L253, I think we have followed the structure of this object in common-types.

As there are other properties (like `None` in the `type` and `userAssignedIdentities`), we are not able to directly refer to the common types.

File: specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json (hunk 3 above)
Comment (jinghzhu-msft) [reply specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json:484]: Hi @raosuhas , the context is there is a bug in Portal side which forces us to use PATCH as a workaround. Here are some discussions with Jeffrey Richter and Arthur Ning about it before: https://github.com/Azure/azure-rest-api-specs-pr/pull/5474#issuecomment-969136518

File: specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json (hunk 2 above)
Comment (jinghzhu-msft) [reply specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json:379]: Have updated it. Please check it again. Thanks.

File: specification/dashboard/resource-manager/readme.typescript.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
+## TypeScript
+
+These settings apply only when `--typescript` is specified on the command line.
+Please also specify `--typescript-sdks-folder=<path to root folder of your azure-sdk-for-js clone>`.
+
+```yaml $(typescript)
+typescript:
+  azure-arm: true
+  package-name: "dashboard"
```
Comment (dw511214992) [inline specification/dashboard/resource-manager/readme.typescript.md:9]:
```suggestion
  package-name: "@azure/arm-dashboard"
```

File: specification/dashboard/resource-manager/readme.csharp.md [hunk 7]
```diff
@@ -0,0 +1,15 @@
+## C
```
Comment (dw511214992) [inline specification/dashboard/resource-manager/readme.csharp.md:1]:
```suggestion
## C#
```

File: specification/dashboard/resource-manager/readme.typescript.md [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  output-folder: "$(typescript-sdks-folder)/packages/dashboard"
```
Comment (dw511214992) [inline specification/dashboard/resource-manager/readme.typescript.md:10]:
```suggestion
  output-folder: "$(typescript-sdks-folder)/sdk/dashboard/arm-dashboard"
```

**Review by dw511214992 — CHANGES_REQUESTED** [review]

File: specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json (hunk 1 above)
Comment (jinghzhu-msft) [reply specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json]: Have updated it.

File: specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json (hunk 3 above)
Comment (jinghzhu-msft) [reply specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json]: After double check, the Portal bug is fixed. I've removed the `location` from PATCH.

File: specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json (hunk 5 above)
Comment (jinghzhu-msft) [reply specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json:612]: Has updated our Swagger spec to be aligned with the ones from common-types. Please check it again.

File: specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json [hunk 9]
```diff
@@ -0,0 +1,760 @@
...
+          "description": "The endpoint of the Grafana instance.",
+          "type": "string",
+          "readOnly": true
+        },
+        "zoneRedundancy": {
+          "$ref": "#/definitions/ZoneRedundancy"
+        }
+      }
+    },
+    "GrafanaResourcePropertiesUpdateParameters": {
```
[the remark this answers is not in the document]
Comment (jinghzhu-msft) [reply specification/dashboard/resource-manager/Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json]: Thanks for the comment. I've removed it.

File: specification/dashboard/resource-manager/readme.md [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+```
+
+### Tag: package-2021-09-01-preview
+
+These settings apply only when `--tag=package-2021-09-01-preview` is specified on the command line.
+
+``` yaml $(tag) == 'package-2021-09-01-preview'
+input-file:
+    - Microsoft.Dashboard/preview/2021-09-01-preview/grafana.json
+```
```
Comment (dw511214992) [inline specification/dashboard/resource-manager/readme.md:36]: Please add a section `Swagger to SDK` in readme.md. An example can be found here: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/readme.md#swagger-to-sdk

File: specification/dashboard/resource-manager/readme.go.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+```
+
+### Tag: package-2021-09-01-preview and go
+
+These settings apply only when `--tag=package-2021-09-01-preview --go` is specified on the command line.
+Please also specify `--go-sdks-folder=<path to the root directory of your azure-sdk-for-go clone>`.
+
+```yaml $(tag) == 'package-2021-09-01-preview' && $(go)
+output-folder: $(go-sdks-folder)/services/preview/$(namespace)/mgmt/2021-09-01-preview/$(namespace)
+```
```
Comment (dw511214992) [inline specification/dashboard/resource-manager/readme.go.md:26]:
~~~suggestion
## Go

These settings apply only when `--go` is specified on the command line.

```yaml $(go) && $(track2)
azure-arm: true
license-header: MICROSOFT_MIT_NO_VERSION
module-name: sdk/resourcemanager/dashboard/armdashboard
module: github.com/Azure/azure-sdk-for-go/$(module-name)
output-folder: $(go-sdk-folder)/$(module-name)
```
~~~

File: specification/dashboard/resource-manager/readme.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+This section describes what SDK should be generated by the automatic system.
+This is not used by Autorest itself.
+
+``` yaml $(swagger-to-sdk)
+swagger-to-sdk:
+  - repo: azure-sdk-for-net
+  - repo: azure-sdk-for-python
+  - repo: azure-sdk-for-python-track2
+  - repo: azure-sdk-for-go
+  - repo: azure-sdk-for-go-track2
```
Comment (dw511214992) [inline specification/dashboard/resource-manager/readme.md:51]:
```suggestion
  - repo: azure-sdk-for-go-track2
```

**Comment by msyyc** [comment]
Hi @jinghzhu-msft , could you tell me where do you get configuration template for `readme.python.md`? We have [template](https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/samplefiles/readme.python.md) but it seems that it is not used.

File: specification/dashboard/resource-manager/readme.csharp.md [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+```yaml $(csharp)
+csharp:
+  azure-arm: true
+  license-header: MICROSOFT_MIT_NO_VERSION
+  payload-flattening-threshold: 1
+  clear-output-folder: true
+  client-side-validation: false
+  namespace: Microsoft.Dashboard
+  output-folder: $(csharp-sdks-folder)/dashboard/management/Microsoft.Dashboard/src/Generated
```
Comment (dw511214992) [inline specification/dashboard/resource-manager/readme.csharp.md:14]:
```suggestion
  namespace: Microsoft.Azure.Management.Dashboard
  output-folder: $(csharp-sdks-folder)/dashboard/Microsoft.Azure.Management.Dashboard/src/Generated
```

**Review by dw511214992 — APPROVED** [review]

**Pull request merged**